addons/coupon/models/coupon_program.py

```diff
@@ -0,0 +1,160 @@
+
+    @api.depends('coupon_ids')
+    def _compute_coupon_count(self):
+        coupon_data = self.env['coupon.coupon'].read_group([('program_id', 'in', self.ids)], ['program_id'], ['program_id'])
+        mapped_data = dict([(m['program_id'][0], m['program_id_count']) for m in coupon_data])
+        for program in self:
+            program.coupon_count = mapped_data.get(program.id, 0)
+
+    @api.onchange('promo_code_usage')
+    def _onchange_promo_code_usage(self):
+        if self.promo_code_usage == 'no_code_needed':
+            self.promo_code = False
+
+    @api.onchange('reward_product_id')
+    def _onchange_reward_product_id(self):
+        if self.reward_product_id:
+            self.reward_product_uom_id = self.reward_product_id.uom_id
+
+    @api.model
+    def create(self, vals):
+        program = super(CouponProgram, self).create(vals)
+        if not vals.get('discount_line_product_id', False):
+            discount_line_product_id = self.env['product.product'].create({
+                'name': program.reward_id.display_name,
+                'type': 'service',
+                'taxes_id': False,
+                'supplier_taxes_id': False,
+                'sale_ok': False,
+                'purchase_ok': False,
+                'lst_price': 0, #Do not set a high value to avoid issue with coupon code
+            })
+            program.write({'discount_line_product_id': discount_line_product_id.id})
+        return program
+
+    def write(self, vals):
+        res = super(CouponProgram, self).write(vals)
+        reward_fields = [
+            'reward_type', 'reward_product_id', 'discount_type', 'discount_percentage',
+            'discount_apply_on', 'discount_specific_product_ids', 'discount_fixed_amount'
+        ]
+        if any(field in reward_fields for field in vals):
+            self.mapped('discount_line_product_id').write({'name': self[0].reward_id.display_name})
+        return res
+
+    def unlink(self):
+        for program in self.filtered(lambda x: x.active):
+            raise UserError(_('You can not delete a program in active state'))
+        return super(CouponProgram, self).unlink()
+
+    def toggle_active(self):
+        super(CouponProgram, self).toggle_active()
+        if not self.active:
+            coupons = self.filtered(lambda p: p.promo_code_usage == 'code_needed').mapped('coupon_ids')
+            coupons.filtered(lambda x: x.state != 'used').write({'state': 'expired'})
+        self.mapped('discount_line_product_id').write({'active': self.active})
+
+    def _is_global_discount_program(self):
+        self.ensure_one()
+        return self.promo_applicability == 'on_current_order' and \
+               self.reward_type == 'discount' and \
+               self.discount_type == 'percentage' and \
+               self.discount_apply_on == 'on_order'
+
+    def _keep_only_most_interesting_auto_applied_global_discount_program(self):
```

Comment:
seems to be only used on orders should it go to -> sale_coupon? (same for _is_global_discount_program)

Reply:
Indeed. I can move these guys in sale_coupon module. I'll do the same to others that are only applicable in sale_coupon.